test/dynamo/test_decorators.py

```diff
@@ -211,6 +211,421 @@ def test_incorrect_usage_disallow_in_graph(self):
+        opt_fn = torch.compile(fn, dynamic=True, fullgraph=True, backend="aot_eager")
+
+        ref = fn(x)
+        res = opt_fn(x)
+        self.assertEqual(ref, res)
+
+    def test_mark_traceable_inside_compiled_function(self):
+        def trace_me(x):
+            torch._dynamo.graph_break()
+            return x + 42
+
+        def fn(x):
+            res = torch._dynamo.mark_traceable(trace_me)(x)
+            return res + 1
+
+        x = torch.randn(10)
+        opt_fn = torch.compile(fn, fullgraph=True, backend="aot_eager")
+
+        ref = fn(x)
+        res = opt_fn(x)
+        self.assertEqual(ref, res)
+
+    def test_mark_traceable_no_action_at_a_distance(self):
+        def trace_me(x):
+            torch._dynamo.graph_break()
+            return x + 42
+
+        # No effect on traceability of `trace_me`
+        torch._dynamo.mark_traceable(trace_me)
+
+        def fn(x):
+            res = trace_me(x)
+            return res + 1
+
+        x = torch.randn(10)
+        cnts = torch._dynamo.testing.CompileCounterWithBackend("aot_eager")
+        opt_fn = torch.compile(fn, backend=cnts)
+
+        ref = fn(x)
+        res = opt_fn(x)
+        self.assertEqual(ref, res)
+        # There should be 1 graph break
+        self.assertEqual(cnts.frame_count, 2)
+
+    def test_mark_traceable_inside_compiled_function_error(self):
+        @torch.compile(fullgraph=True, backend="aot_eager")
+        def fn(x, y):
+            def trace_me(x, y):
+                torch._dynamo.graph_break()
+                return x * y
+
+            res = torch._dynamo.mark_traceable(trace_me)(x, y)
+            return res + 1
+
+        try:
+            fn(torch.ones(10), torch.ones(1))
+            self.assertFalse(True)  # must raise error before this
+        except torch._dynamo.exc.Unsupported as e:
+            msg = """
+`mark_traceable` currently requires the target function to be defined outside `torch.compile` region.
+"""  # NOQA: B950
+            self.assertIn(msg, str(e))
+
+    def test_mark_traceable_custom_class_error(self):
+        class Point:
+            x: torch.Tensor
+            y: torch.Tensor
+
+            def __init__(self, x, y):
+                self.x = x
+                self.y = y
+
+        @torch._dynamo.mark_traceable
+        def trace_me(p):
+            torch._dynamo.graph_break()
+            return p.x * p.y
+
+        @torch.compile(fullgraph=True, backend="aot_eager")
+        def fn(p):
+            res = trace_me(p)
+            return res + 1
+
+        try:
+            p = Point(torch.ones(10), torch.ones(1))
+            fn(p)
+            self.assertFalse(True)  # must raise error before this
+        except torch._dynamo.exc.Unsupported as e:
+            msg = """
+Attempting to call a `mark_traceable`-ed function with arguments that contain a value of type <DecoratorTests.test_mark_traceable_custom_class_error.<locals>.Point>, please use one of the following to register the type with pytree:
```

Comment:
Error message should have something like "The only allowed input/output types are basic types (like Tensors, ints, floats, <fill me in here>) or pytree containers of those. If your class acts like a pytree container, then please use one of the following to register...")

The "acts like a pytree container" part is important - aka the object identity doesn't matter.

Reply:
Will update, but I thought showing the class name will help user figure out which type to pytree-register, especially when there are lots of types involved in a nested data structure?

Reply:
Having the class name is good. I wanted to point out that the user might not know that mark_traceable requires pytree inputs so it's good to point it out